js/stanford_paragraph_types.p_hero.js

```diff
@@ -3,44 +3,131 @@
  * Applies some functionality on the paragraph types.
  */
 
+/**
+ * Called automatically by YouTube's API.
+ */
+function onYouTubeIframeAPIReady() {
+  // After Youtube API is loaded, we can do what we need.
+  Drupal.behaviors.stanfordParagraphPHero.youTubeReady()
+}
+
 (function ($) {
   Drupal.behaviors.stanfordParagraphPHero = {
+    heroFunctions: this,
+
     attach: function (context, settings) {
-      $('.field-name-field-p-hero-image', context).each(function () {
-        var video = $(this).siblings('.field-name-field-p-hero-video');
-
-        if (video.length) {
-          $(video).find('iframe').attr('title', Drupal.t('Video Player'));
-          var videoUrl = $(video).find('div[data-video-embed-url]').attr('data-video-embed-url');
-
-          var play = $('<a>', {
-            class: 'play-video',
-            href: videoUrl,
-            html: $('<i>', {
-              class: 'fa fa-youtube-play icon-youtube-play',
-              html: 'Play Video',
-              'aria-label': Drupal.t('Play Video - Opens to the video website')
-            })
-          }).click(function (e) {
-
-            // Mouse has eventPhase 3, keyboard has 2.
-            if (e.eventPhase == 3) {
-              e.preventDefault();
-              $dad = $(this).parent();
-              $dad.hide();
-              $dad.siblings('.group-overlay-text').hide();
-              var iframe = $(video).find('iframe')[0];
-
-              iframe.src += "&autoplay=1";
-              $(iframe).attr('onload', 'this.contentWindow.focus()');
-              $(video).show();
-            }
-
-          });
-
-          $(this).prepend(play);
+      var i = 0;
+      $('.field-name-field-p-hero-video iframe').each(function () {
+        // Make sure the video has an ID attribute.
+        if (typeof $(this).attr('id') == 'undefined') {
+          $(this).attr('id', 'hero-video-' + i);
         }
+        i++;
+      });
+
+
+      var tag = document.createElement('script');
+      var firstScriptTag = document.getElementsByTagName('script')[0];
+
+      if ($('.field-name-field-p-hero-video iframe[src*="youtube"]').length) {
+        tag.src = '//www.youtube.com/iframe_api';
+        firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);
+      }
+
+      if ($('.field-name-field-p-hero-video iframe[src*="vimeo"]').length) {
+        $.getScript('//player.vimeo.com/api/player.js', function () {
+          Drupal.behaviors.stanfordParagraphPHero.vimeoReady();
+        });
+      }
+
+    },
+
+    youTubeReady: function () {
+      $heroFunctions = this;
```

Comment:
This line is very similar to line 16 but is completely different. `$heroFunctions` is not the same variable as `heroFunctions` and they do get set to different things. One gets set to `Drupal.behaviors.stanfordParagraphPHero` and the other gets set to `Drupal.behaviors.stanfordParagraphPHero.youTubeReady`. Please clear up the naming of these variables.

Reply:
Also, it is not clear what they are used for. Please add comments to them.